inject/src/main/java/io/micronaut/context/DefaultBeanContext.java

```diff
@@ -2193,12 +2193,37 @@ protected <T> Collection<BeanDefinition<T>> findBeanCandidates(@Nullable BeanRes
      * @param predicate         The predicate to filter candidates
      * @return The candidates
      */
-    @SuppressWarnings("unchecked")
     @NonNull
     protected <T> Collection<BeanDefinition<T>> findBeanCandidates(@Nullable BeanResolutionContext resolutionContext,
                                                                    @NonNull Argument<T> beanType,
                                                                    boolean filterProxied,
                                                                    Predicate<BeanDefinition<T>> predicate) {
+        return findBeanCandidates(
+            resolutionContext,
+            beanType,
+            filterProxied,
+            true,
+            predicate
+        );
+    }
+
+    /**
+     * Find bean candidates for the given type.
+     *
+     * @param <T>               The bean generic type
+     * @param resolutionContext The current resolution context
+     * @param beanType          The bean type
+     * @param filterProxied     Whether to filter out bean proxy targets
+     * @param collectIterables  Whether iterables should be collected
+     * @param predicate         The predicate to filter candidates
+     * @return The candidates
+     */
+    @NonNull
+    protected <T> Collection<BeanDefinition<T>> findBeanCandidates(@Nullable BeanResolutionContext resolutionContext,
```

Comment:
I think we should refactor out all the caching and logic around `BeanDefinition` resolution into a separate class and use it as a field to reduce the complexity of the class, `DefaultApplicationContext` would have a custom impl. adding transform logic.

Reply:
Agreed, but that can be done in a separate PR